router/widen-lemma.gobra

```diff
@@ -145,828 +153,304 @@ func absPktWidenLemma(dp io.DataPlaneSpec, raw []byte, length int) {
-		hopFieldsConsDirWidenLemma(raw, offset, currHFIdx + 1, (beta union set[io.IO_msgterm]{hf1.HVF}), asid, ainfo, length, newP)
-		ret1 := seq[io.IO_HF]{hf1} ++ hopFieldsConsDir(raw, offset, currHFIdx + 1, (beta union set[io.IO_msgterm]{hf1.HVF}), asid, ainfo)
-		ret2 := seq[io.IO_HF]{hf2} ++ hopFieldsConsDir(raw[:length], offset, currHFIdx + 1, (beta union set[io.IO_msgterm]{hf2.HVF}), asid, ainfo)
-		assert ret1 == hopFieldsConsDir(raw, offset, currHFIdx, beta, asid, ainfo)
-		assert ret2 == hopFieldsConsDir(raw[:length], offset, currHFIdx, beta, asid, ainfo)
-		assert ret1 == ret2
-	}
-}
-
-ghost
-requires  0 <= length && length <= len(raw)
-requires  idx + path.HopLen <= length
-requires  0 <= idx
+requires  0 <= middle
+requires  middle + path.HopLen <= length
+requires  length <= len(raw)
 preserves acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), R54)
 preserves acc(sl.AbsSlice_Bytes(raw[:length], 0, len(raw[:length])), R54)
-ensures   hopField(raw, idx, beta, asid, ainfo) == hopField(raw[:length], idx, beta, asid, ainfo)
+ensures   path.BytesToIO_HF(raw, 0, middle, len(raw)) == path.BytesToIO_HF(raw[:length], 0, middle, length)
 decreases
-func hopFieldWidenLemma(raw []byte, idx int, beta set[io.IO_msgterm], asid io.IO_as, ainfo io.IO_ainfo, length int) {
+func BytesToIO_HFWidenLemma(raw []byte, middle int, length int) {
 	unfold acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), R55)
 	unfold acc(sl.AbsSlice_Bytes(raw[:length], 0, len(raw[:length])), R55)
-
-	assert forall i int :: { &raw[idx+2+i] } { &raw[idx+2:idx+4][i] } 0 <= i && i < 2 ==> &raw[idx+2+i] == &raw[idx+2:idx+4][i]
-	assert forall i int :: { &raw[idx+4+i] } { &raw[idx+4:idx+6][i] } 0 <= i && i < 2 ==> &raw[idx+4+i] == &raw[idx+4:idx+6][i]
-	inif21 := binary.BigEndian.Uint16(raw[idx+2:idx+4])
-	inif22 := binary.BigEndian.Uint16(raw[:length][idx+2:idx+4])
-	assert inif21 == inif22
-
-	egif2 := binary.BigEndian.Uint16(raw[idx+4:idx+6])
-	op_inif2 := inif21 == 0 ? none[io.IO_ifs] : some(io.IO_ifs(inif21))
-	op_egif2 := egif2 == 0 ? none[io.IO_ifs] : some(io.IO_ifs(egif2))
-	ts := io.IO_msgterm(io.MsgTerm_Num{ainfo})
-	l := io.IO_msgterm(io.MsgTerm_L{seq[io.IO_msgterm]{ts, io.if2term(op_inif2), io.if2term(op_egif2), io.IO_msgterm(io.MsgTerm_FS{beta})}})
-	hvf := io.mac(io.macKey(io.asidToKey(asid)), l)
-
-	ret1 := io.IO_HF(io.IO_HF_{
-		InIF2 : op_inif2,
-		EgIF2 : op_egif2,
-		HVF : hvf,
-	})
-	ret2 := io.IO_HF(io.IO_HF_{
-		InIF2 : op_inif2,
-		EgIF2 : op_egif2,
-		HVF : hvf,
-	})
-	assert ret1 == hopField(raw, idx, beta, asid, ainfo)
-	assert ret2 == hopField(raw[:length], idx, beta, asid, ainfo)
-	assert ret1 == ret2
+	assert path.BytesToIO_HF(raw, 0, middle, len(raw)).EgIF2 == path.BytesToIO_HF(raw[:length], 0, middle, length).EgIF2
+	assert path.BytesToIO_HF(raw, 0, middle, len(raw)).InIF2 == path.BytesToIO_HF(raw[:length], 0, middle, length).InIF2
+	assert path.BytesToIO_HF(raw, 0, middle, len(raw)).HVF == path.BytesToIO_HF(raw[:length], 0, middle, length).HVF
```

Comment:
we can introduce vars here for repeated expressions like `path.BytesToIO_HF(raw, 0, middle, len(raw))`